File: app/src/test/java/com/neocaptainnemo/testing/service/AtmsRepoTest.java
package com.neocaptainnemo.testing.service;

import com.google.gson.Gson;
import com.neocaptainnemo.testing.model.ViewPort;

import org.assertj.core.api.Assertions;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.io.InputStream;
import java.io.InputStreamReader;

import rx.Observable;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class AtmsRepoTest {

    @Mock
    OpenStreetMap openStreetMap;

    AtmsRepo repo;

    @Before
    public void before() {
        repo = new AtmsRepo(openStreetMap);
    }

    @Test
    public void fetchSuccessNoNull() {

        InputStream stream = ClassLoader.getSystemResourceAsStream("response.json");
        InputStreamReader reader = new InputStreamReader(stream);
        OsmResponse response = new Gson().fromJson(reader, OsmResponse.class);

        when(openStreetMap.request(any())).thenReturn(Observable.just(response));

        Assertions.assertThat(repo.request(new ViewPort(1, 1, 1, 1)).toBlocking().first())
                .isEqualTo(response.getAtms());

    }

    @Test
    public void fetchNull() {

        InputStream stream = ClassLoader.getSystemResourceAsStream("response_null.json");
        InputStreamReader reader = new InputStreamReader(stream);
        OsmResponse response = new Gson().fromJson(reader, OsmResponse.class);

        when(openStreetMap.request(any())).thenReturn(Observable.just(response));

        Assertions.assertThat(repo.request(new ViewPort(1, 1, 1, 1)).toBlocking().first())
                .isEmpty();
    }

}

File: app/src/test/java/com/neocaptainnemo/atms/service/AtmsRepoTest.kt
package com.neocaptainnemo.atms.service

import com.google.gson.Gson
import com.neocaptainnemo.atms.model.ViewPort
import com.nhaarman.mockito_kotlin.whenever
import io.reactivex.Observable
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mock
import org.mockito.junit.MockitoJUnitRunner
import java.io.InputStreamReader

@RunWith(MockitoJUnitRunner::class)
class AtmsRepoTest {

    @Mock
    lateinit var openStreetMap: OpenStreetMap

    lateinit var repo: AtmsRepo

    @Before
    fun before() {
        repo = AtmsRepo(openStreetMap)
    }

    @Test
    fun fetchSuccessNoNull() {

        val stream = ClassLoader.getSystemResourceAsStream("response.json")
        val reader = InputStreamReader(stream)
        val response = Gson().fromJson(reader, OsmResponse::class.java)

        whenever(openStreetMap.request(anyString())).then { Observable.just(response) }

        val testable = repo.request(ViewPort(1.0, 1.0, 1.0, 1.0)).test()

        testable.assertValue(response.atms)
    }

    @Test
    fun fetchNull() {

        val stream = ClassLoader.getSystemResourceAsStream("response_null.json")
        val reader = InputStreamReader(stream)
        val response = Gson().fromJson(reader, OsmResponse::class.java)

        whenever(openStreetMap.request(anyString())).then { Observable.just(response) }

        val testable = repo.request(ViewPort(1.0, 1.0, 1.0, 1.0)).test()

        testable.assertValue { it.isEmpty() }
    }

}
